### ISSUE TITLE
Feature Request: Improved rollout strategy

### ISSUE THREAD

**Issue by dndungu** [keikoproj/upgrade-manager#45]
## Challenges
The current rolling upgrade strategy presents the following challenges.
1. Only one instance per AZ can be safely terminated since the strategy is to terminate first, then wait for the ASG to replace terminated instances.
2. Pods can get rescheduled several times during an upgrade. There is no guarantee that pods will get scheduled in the instances using the new Launch Configuration.

## Proposal
1. Create a new ASG with min=0, desired=**n** (some configurable value), max=(same value as old ASG)
1. Wait for the **n** instances to get created and be healthy nodes in the Cluster.
1. Cordon all the instances in the old ASG to prevent the pods from being scheduled in them since we already know they will soon be terminated.
1. Drain **n** instances in the old ASG.
1. Wait for the pods to get successfully scheduled on the new node.
1. For every node successfully drained and until new ASG desired==old ASG desired:
    1. Terminate the drained node.
    1. Increase the desired instances in the new ASG
    1. Wait for the new instance to be created and healthy
    1. Drain another node in the old ASG
    1. Wait for pods to get successfully scheduled in the new pod
1. Delete the old ASG

**Comment by shrinandj** [comment]
@dndungu Thanks for filing this issue. 

The strategy that you mentioned (kinda like blue-green upgrades) had been discussed before. The problem usually is for large ASGs with lots of nodes. In those cases, it may not be possible to spin up **n** new nodes because the AWS account does not have enough instances of that type. Ideally, the code should handle this by simply accepting whatever instances it got and keep the upgrade rolling with those many instances in each "batch". The safest and most reliable approach was to terminate instance 1 by 1 and that's why it was the first one implemented.

There are some more strategies wherein multiple instances are terminated in one go.

The strategy that you propose is certainly efficient and definitely doable. It just needs to handle the corner cases better.

PRs welcome ;-)

**Comment by eytan-avisror** [comment]
In addition, this might be problematic with things like `Kops` which need to be the provisioner of scaling groups, this will be very hard to implement generically, where when we just kill instances we kinda keep it simple.

**Comment by dndungu** [comment]
@shrinandj, thanks for the response, we have faced the AWS capacity issues a number of times. 

A possible solution could be to use create a capacity reservation before starting the rollout and cancel it after a successful rollout. This could also be coupled with having fallback instance types to be used when AWS does not have the desired instance type available.

**Comment by uthark** [comment]
> The safest and most reliable approach was to terminate instance 1 by 1 and that's why it was the first one implemented.

I'd also try to go opposite - start new instance first and then terminate old one.

**Comment by shrinandj** [comment]
| I'd also try to go opposite - start new instance first and then terminate old one.

@kianjones4 changed the AWS api for doing something similar. Instead of simply calling ec2.TerminateInstances, upgrade-manager now calls autoscaling.TerminateInstancesInAutoScalingGroup that spins up a new node first and then terminates the existing one. But if there are other improvements to be made here, feel free to file a separate ticket.

As for this specific ticket, I *think*, we might be able to use the new AWS vCPU based quotas to our advantage here. AWS recently changed the way they account for instances per account. Earlier, there was a limit per instance type (e.g. 100 instances of m5.large, 40 instances of m4.xlarge, etc.). Now, the total limit for all `m` instances is say 1000vCPUs. How these are divided between m5.large, xlarge, etc. does not matter. There is also an API for querying the current vCPU limit.

So, the upgrade strategy proposed here could actually be implemented nicely. Here are the steps.

- Get the number of instances in the ASG. Based on the launchConfig/launchTemplate, calculate the number of vCPUs being used.
- Query AWS to check if the total quota for the instances in question is at least double the number found in the above step.
- If yes, simply double the size of the ASG. Wait for the instances to come up, drain and terminate the old instances.
- If no, either error out or fall back to the default strategy.

What do you'll think?

**Comment by dndungu** [comment]
@shrinandj `TerminateInstancesInAutoScalingGroup` definitely changes things. We can essentially do a blue-green upgrade provided the instances only get  marked healthy after joining the cluster.

**Comment by edlee2121** [comment]
Yes, I also think it is preferable to start new instances before terminating the old. Perhaps we can also prefer scheduling on the new nodes vs. the old to minimize the number of pod evictions.

**Comment by eytan-avisror** [comment]
The problem with having instances join before they terminate is that you will have to play around with the min/max/desired.
e.g. bump up max, bump up desired -> kill old instance -> revert to old settings.
This adds a lot of state that needs to be persisted somewhere in case you crash after modifying these settings.

Right now, with using `TerminateInstancesInAutoScalingGroup` the behavior is that when an instance moves to `Terminating:Proceed` (in case you are using hooks), a new instance is launched to immediately, so as the instance is about to terminate, a new one is already pending.

If we can think of a way to launch a new instance before termination without touching the ASG config this would be ideal, but I don't think this is possible as far as I am aware.

Considering all different kinds of provisioners which may be in use out there, modifying the ASG might not be appropriate

**Comment by eytan-avisror** [comment]
We should evaluate detaching an instance from an ASG, this will mean ASG will do the right thing and launch a new one while we still have the old instance around.
But this complicates things with lifecycle-hooks as you will probably never get one in this case.

**Comment by eytan-avisror** [comment]
I spoke too soon, here is how we can achieve this:
https://docs.aws.amazon.com/sdk-for-go/api/service/autoscaling/#EnterStandbyInput

If you set an instance to Standby mode, you can ask ASG to spin up a new instance in it's place.
Once instance is joined, call `TerminateInstancesInAutoScalingGroup` on that standby instance.

We should do this ! :)

**Comment by eytan-avisror** [comment]
So the ordering of actions can be:

- Drain/Cordon "Old" instance
- Set it to standby mode (no decrement capacity)
- Wait for "New" instance to be in service / join cluster (need to think about this readiness condition)
- `TerminateInstancesInAutoScalingGroup` => "Old" instance
- Lifecycle hooks triggered on "Old" instance

No need to touch ASG at all in this case, only instances.

**Comment by eytan-avisror** [comment]
I did some preliminary work on this, I believe this will solve for at-least some of the concerns raised here.
Feel free to try it out, I will add some unit tests later so that we can potentially merge it.

**Comment by eytan-avisror** [comment]
@dndungu @uthark 
This is now merged, any feedback appreciated.

**Comment by dndungu** [comment]
Thank you @eytan-avisror, for the effort you have put in to this. I believe you have addressed the biggest pain point here.

**Comment by eytan-avisror** [comment]
Thanks @dndungu, that's great! I will go ahead a close this

**Comment by shrinandj** [comment]
Thanks guys! Upgrade-manager v0.6 has been released with support for this feature.

Release: https://github.com/keikoproj/upgrade-manager/releases/tag/v0.6
Docker image: `keikoproj/rolling-upgrade-controller:0.6`
